jetbrains-core/src/software/aws/toolkits/jetbrains/ui/AsyncComboBox.kt

```
@@ -0,0 +1,109 @@
+
+                if (loading.get() && index == -1) {
+                    component.icon = AnimatedIcon.Default.INSTANCE
+                    component.text = "Loading"
+                }
+
+                return component
+            }
+
+            override fun customize(list: JList<out T>, value: T, index: Int, selected: Boolean, hasFocus: Boolean) {
+                customizer?.customize(this, value, index)
+            }
+        }
+    }
+
+    private val reloadTimer = Timer()
+    private var reloadTimerTask: TimerTask? = null
+
+    @Synchronized
+    fun proposeModelUpdate(modelConsumer: suspend (DefaultComboBoxModel<T>) -> Unit) {
+        loading.set(true)
+        removeAllItems()
+        reloadTimerTask?.cancel()
+        // debounce
+        reloadTimerTask = timerTask {
+            val event = ListDataEvent(this, ListDataEvent.CONTENTS_CHANGED, -1, -1)
+            // force redraw of renderer without actually touching the underlying model
+            comboBoxModel.listDataListeners?.forEach { it.contentsChanged(event) }
+            runBlocking(scope.coroutineContext) {
+                try {
+                    modelConsumer.invoke(comboBoxModel)
+                } finally {
+                    loading.set(false)
+                }
+            }
+            comboBoxModel.listDataListeners?.forEach { it.contentsChanged(event) }
+        }
+        reloadTimer.schedule(reloadTimerTask, 350)
+    }
+
+    override fun dispose() {
+    }
+
+    override fun getSelectedItem(): Any? {
+        if (loading.get()) {
+            return null
+        }
+        return super.getSelectedItem()
+    }
+
+    @TestOnly
+    @Synchronized
+    internal fun blockingGet(duration: Duration): T? {
+        try {
+            spinUntil(duration) { !loading.get() }
+        } catch (e: Exception) {
+            return null
+        }
```

Should probably let this throw, so we can test null cases safely.